app/LDAP/LDAPFunctions.php

```diff
@@ -0,0 +1,464 @@
+	 */
+	protected function LDAP_set_option(int $opt, string $value): void
+	{
+		try {
+			\Safe\ldap_set_option($this->con, $opt, $value);
+		} catch (\Throwable $e) {
+			throw new LDAPException($e->getMessage(), $e);
+		}
+	}
+
+	/**
+	 * Warp around ldap_start_tls.
+	 *
+	 * @return void
+	 *
+	 * @throws LDAPException
+	 */
+	protected function LDAP_start_tls(): void
+	{
+		try {
+			if (!ldap_start_tls($this->con)) {
+				throw new \Safe\Exceptions\LdapException('ldap_stat_tls failed');
+			}
+		} catch (\Throwable $e) {
+			throw new LDAPException($e->getMessage(), $e);
+		}
+	}
+
+	/**
+	 * Reads the user data from the LDAP server.
+	 *
+	 * @param string $username
+	 *
+	 * @return LDAPUserData contains the user data
+	 *
+	 * @throws LDAPException
+	 */
+	public function get_user_data(string $username): LDAPUserData
+	{
+		$this->open_LDAP();
+
+		if (!empty($this->cached_user_info) && in_array($username, $this->cached_user_info)) {
+			Logs::notice(__METHOD__, __LINE__, sprintf('getUserData: Use cached info for %s', $username));
+
+			return $this->cached_user_info[$username];
+		}
+
+		// force superuser bind if wanted and not bound as superuser yet
+		if (Configs::get_value(self::CONFIG_KEY_BIND_DN) && Configs::get_value(self::CONFIG_KEY_BIND_PW) && $this->bound < self::BIND_TYPE_SUPER_USER) {
+			// use superuser credentials
+			$this->LDAP_bind();
+		}
+
+		$userData = new LDAPUserData();
+		$userData->user = $username;
+
+		// get info for given user
+		$base = self::_makeFilter(Configs::get_value(self::CONFIG_KEY_USER_TREE), $userData->toArray());
+		Logs::notice(__METHOD__, __LINE__, sprintf('base filter: %s', $base));
+		if (Configs::get_value(self::CONFIG_KEY_USER_FILTER)) {
+			$filter = self::_makeFilter(Configs::get_value(self::CONFIG_KEY_USER_FILTER), $userData->toArray());
+		} else {
+			$filter = '(ObjectClass=*)';
+		}
+		Logs::notice(__METHOD__, __LINE__, sprintf('filter: %s', $filter));
+
+		// TODO fix: This line will fail; this is suspicious.
+		//
+		// This line will fail if this method is invoked independently of
+		// {@link self::check_pass()} and no super-user is configured which
+		// has bound to the LDAP server some-lines above.
+		//
+		// The only reason why this works is that "accidentally"
+		// `check_pass` and `get_user_data` are invoked in that exact order
+		// in `SessionFunctions` and hence the previously called
+		// `check_pass` has taken care of proper binding.
+		//
+		// Either we intent to always call `check_pass` and `get_user_data`
+		// in that order.
+		// In that case `get_user_data` should not be a public function but
+		// the result of check user data should be returned as part of
+		// `check_pass`.
+		//
+		// Or, if we intent to use `get_user_data` stand-alone, then this
+		// must be fixed in a way such that it works independent of a
+		// previous call to `check_pass`.
+		$result = $this->LDAP_search($base, $filter, Configs::get_value(self::CONFIG_KEY_USER_SCOPE));
```

Comment:
While I have been working on it, I had to understand the code and found this error in the logic of the program flow. This is your to-do. Please take care of it.